### DISCUSSION TITLE
Possible to implement something like Haskell's syntax for writing arbitrary / custom infix functions in user space?

### DISCUSSION THREAD

**Discussion by ym-han** [eclipse-langium/langium#1675] [Q&A] [answered]
Haskell allows users to define custom infix operators (see, e.g., https://jplahn.gitbooks.io/a-happy-haskell/content/functions/infix_operators.html for examples).

Is it possible to make a Langium grammar, or extend the Langium parser, to similarly allow users to declare / define custom infix notation? (or to do perhaps a nearby but weaker variant of this?)

**Comment by msujew** [comment, marked as answer]
Hey @ym-han,

I think that should be possible purely by implementing additional parser rules. Kind of similar to the arithmetics example, you would have a new kind of expression:

https://github.com/eclipse-langium/langium/blob/b2a6fe9684434117eb6a344fc2721e3c26d989bf/examples/arithmetics/src/language-server/arithmetics.langium#L25-L26

```
InfixFunctionCallExpression infers Expression: 
  Multiplication 
  ({infer InfixFunctionCallExpression.left=current} 
    // The "operator" is actually a reference to a function
    operator=[Function:BacktickString] 
    right=Multiplication
  )*;

terminal BacktickString: /`[^`]*`/;
```

Note that you can replace `BacktickString` with a different terminal, depending on your language. For example, AFAIK haskell also allows something like this:

```hs
<+> :: Num a => a -> a -> a
<+> b = a + b

ghci> 1 <+> 1
2
```

That would require something like this:

```
operator=[Function:InfixOperatorText] 
...

InfixOperatorText returns string: AnyOperator | BacktickString;

terminal AnyOperator: /(\!#\$%&\*\+\.\/<=>\?@\\^\|-~)+/;
terminal BacktickString: /`[^`]*`/;
```

**Reply by ym-han** [reply]
Very nice, thank you!!!!

I'm guessing things like fixity declarations are not going to be possible, but most DSLs probably don't require that
